File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        
        int srow=0, scol=0, lrow=n-1, lcol=m-1;
        int cnt = 0;
        List<Integer> ans = new ArrayList<>();
        while(cnt < m*n){
            //if(cnt > m*n) break;
            //first row -> traverse on columns
            for(int j=scol; j<=lcol; j++){
                ans.add(mat[srow][j]);
                cnt++;
            }
            srow++; // 1
            if(cnt >= m*n) break;
            //last col -> t........ on rows
            for(int i=srow; i<=lrow; i++){
                ans.add(mat[i][lcol]);
                cnt++;
            }
            lcol--;   // 3
            if(cnt >= m*n) break;
            //last row
            for(int j=lcol; j>=scol; j--){
                ans.add(mat[lrow][j]);
                cnt++;
            }
            lrow--;  // 3
            if(cnt >= m*n) break;
            //first col
            for(int i=lrow; i>=srow; i--){
                ans.add(mat[i][scol]);
                cnt++; 
            }
            scol++;  // 1
            if(cnt >= m*n) break;
        }
        
        return ans;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> ans;
        
        int rows = 0, rowe = m - 1, cols = 0, cole = n - 1;
        
        while (rows <= rowe && cols <= cole) {
            // Traverse from left to right along the top row
            for (int col = cols; col <= cole; col++) {
                ans.push_back(matrix[rows][col]);
            }
            rows++;
            
            // Traverse from top to bottom along the right column
            for (int row = rows; row <= rowe; row++) {
                ans.push_back(matrix[row][cole]);
            }
            cole--;
            
            // Traverse from right to left along the bottom row, if rows is still within bounds
            if (rows <= rowe) {
                for (int col = cole; col >= cols; col--) {
                    ans.push_back(matrix[rowe][col]);
                }
                rowe--;
            }
            
            // Traverse from bottom to top along the left column, if cols is still within bounds
            if (cols <= cole) {
                for (int row = rowe; row >= rows; row--) {
                    ans.push_back(matrix[row][cols]);
                }
                cols++;
            }
        }
        
        return ans;
    }
};
